scipy/stats/tests/test_sampling.py

```diff
@@ -0,0 +1,380 @@
+            RuntimeError, match=r"c < -0.5 not implemented yet"
+        ):
+            TransformedDensityRejection(
+                lambda x: x, lambda x: 1, domain=(0, 10), c=-np.inf
+            )
+        #  c > 0
+        with pytest.warns(UserWarning, match=r"c > 0"):
+            TransformedDensityRejection(
+                lambda x: x, lambda x: 1, domain=(0, 10), c=10.0
+            )
+        # c = nan
+        with pytest.raises(ValueError, match=r"must be a non-nan value"):
+            TransformedDensityRejection(
+                lambda x: x, lambda x: 1, domain=(0, 10), c=np.nan
+            )
+
+    def test_bad_variant(self):
+        with pytest.raises(
+            ValueError, match=r"Invalid option for the `variant`"
+        ):
+            TransformedDensityRejection(
+                lambda x: x, lambda x: 1, domain=(0, 10), variant="foo"
+            )
+
+    # TODO: test other parameters
+
+
+class TestDiscreteAliasUrn:
+    @pytest.mark.parametrize(
+        "dist, params",
+        [  # discrete distributions with finite support.
+            (stats.hypergeom, (20, 7, 12)),
+            (stats.nhypergeom, (20, 7, 12)),
+            (stats.binom, (20, 0.3)),
+        ],
+    )
+    def test_sampling_with_pmf(self, dist, params):
+        domain = dist.support(*params)
+        pmf = dist._pmf
+        with suppress_warnings() as sup:
+            sup.filter(UserWarning)
+            rng = DiscreteAliasUrn(
+                pmf, domain=domain, params=params, seed=123
+            )
+        rvs = rng.rvs(100_000)
+        # test if the first few moments match
+        mv = rvs.mean(), rvs.var()
+        mv_expected = dist.stats(*params, moments="mv")
+        assert_allclose(mv, mv_expected, atol=1e-1)
+
+        pv = pmf(np.arange(domain[0], domain[1] + 1), *params)
+        # correct for some numerical errors
+        pv = pv / pv.sum()
+        obs_freqs = np.zeros_like(pv)
+        # chi-squared test for goodness-of-fit
+        _, freqs = np.unique(rvs, return_counts=True)
+        freqs = freqs / freqs.sum()
+        obs_freqs[:freqs.size] = freqs
+        with suppress_warnings() as sup:
+            sup.filter(RuntimeWarning, "divide by zero encountered in "
+                                       "true_divide")
+            sup.filter(RuntimeWarning, "invalid value encountered in "
+                                       "true_divide")
+            pval = chisquare(obs_freqs, pv).pvalue
+        assert_allclose(pval, 1.0, atol=1e-2)
+
+    @pytest.mark.parametrize(
+        "bad_pmf, msg",
+        [
+            (lambda x: foo, r"name 'foo' is not defined"),
+            (
+                lambda x, a, b: x + a + b,
+                r"missing 2 required positional arguments: 'a' and 'b'",
+            ),
+            (lambda x: -x, r"50 : probability < 0"),
+            (lambda x: None, r"must be real number, not NoneType"),
+            (lambda x: np.inf, r"240 : unknown error"),
```

Comment:
> we could think about a simple check to sum the pmf values: if that does not give 1

I think the sum does not need to be 1. UNU.RAN normalizes the values after creating the PV from the PMF. 

> for inf, nan and 0, the error message is not very informative :)

Unfortunately, for discrete distributions, float `inf` and `nan` aren't recognized by UNU.RAN and it gives `unknown error`.

> this should cover nan, inf, < 0

We can also just evaluate the PMF at each point in the domain, check if it produces `inf`, `nan`, or `<0` values and if it does raise an error in `_validate_args` otherwise pass the calculated PV to the method. This will help get rid of a warning message and also `unknown error`. What do you think?

Reply:
Sounds like a good idea

Reply:
Done.